doc/source/user/basics.interoperability.rst

```diff
@@ -0,0 +1,352 @@
+
+***************************
+Interoperability with NumPy
+***************************
+
+NumPy’s ndarray objects provide both a high-level API for operations on
+array-structured data and a concrete implementation of the API based on
+`strided in-RAM storage <https://numpy.org/doc/stable/reference/arrays.html>`__.
+While this API is powerful and fairly general, its concrete implementation has
+limitations. As datasets grow and NumPy becomes used in a variety of new
+environments and architectures, there are cases where the strided in-RAM storage
+strategy is inappropriate, which has caused different libraries to reimplement
+this API for their own uses. This includes GPU arrays (CuPy_), Sparse arrays
+(`scipy.sparse`, `PyData/Sparse <Sparse_>`_) and parallel arrays (Dask_ arrays)
+as well as various NumPy-like implementations in deep learning frameworks, like
+TensorFlow_ and PyTorch_. Similarly, there are many projects that build on top
+of the NumPy API for labeled and indexed arrays (XArray_), automatic
+differentiation (JAX_), masked arrays (`numpy.ma`), physical units
+(astropy.units_, pint_, unyt_), among others that add additional functionality
+on top of the NumPy API.
+
+Yet, users still want to work with these arrays using the familiar NumPy API and
+re-use existing code with minimal (ideally zero) porting overhead. With this
+goal in mind, various protocols are defined for implementations of
+multi-dimensional arrays with high-level APIs matching NumPy.
+
+Using arbitrary objects in NumPy
+--------------------------------
+
+When NumPy functions encounter a foreign object, they will try (in order):
+
+1. The buffer protocol, described `in the Python C-API documentation
+   <https://docs.python.org/3/c-api/buffer.html>`__.
+2. The ``__array_interface__`` protocol, described
+   :ref:`in this page <arrays.interface>`. A precursor to Python’s buffer
+   protocol, it defines a way to access the contents of a NumPy array from other
+   C extensions.
+3. The ``__array__`` protocol, which asks an arbitrary object to convert itself
+   into an array.
+
+For both the buffer and the ``__array_interface__`` protocols, the object
+describes its memory layout and NumPy does everything else (zero-copy if
+possible). If that’s not possible, the object itself is responsible for
+returning a ``ndarray`` from ``__array__()``.
+
+The array interface
```

Comment:
We should rename this here and in the main docs to " The `__array_interface__` protocol". Naming it _The array interface_ doesn't really make much sense.